test/e2e/cluster_upgrade.go

```diff
@@ -0,0 +1,158 @@
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/test/framework/kubetest"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+// ClusterUpgradeConformanceSpecInput is the input for ClusterUpgradeConformanceSpec.
+type ClusterUpgradeConformanceSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+}
+
+//  ClusterUpgradeConformanceSpec implements a spec that upgrades a cluster and runs Kubernetes conformance suite.
+// Upgrading a cluster refers to upgrading the control-plane and worker nodes (managed by MD and machine pools)
+func  ClusterUpgradeConformanceSpec(ctx context.Context, inputGetter func() ClusterUpgradeConformanceSpecInput) {
+	const (
+		kubetestConfigurationVariable = "KUBETEST_CONFIGURATION"
+	)
+	var (
+		specName               = "k8s-upgrade-and-conformance"
+		input                  ClusterUpgradeConformanceSpecInput
+		namespace              *corev1.Namespace
+		cancelWatches          context.CancelFunc
+		clusterResources       *clusterctl.ApplyClusterTemplateAndWaitResult
+		kubetestConfigFilePath string
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeFrom))
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeTo))
+		Expect(input.E2EConfig.Variables).To(HaveKey(kubetestConfigurationVariable), "% spec requires a %s variable to be defined in the config file", specName, kubetestConfigurationVariable)
+		kubetestConfigFilePath = input.E2EConfig.GetVariable(kubetestConfigurationVariable)
+		Expect(kubetestConfigFilePath).To(BeAnExistingFile(), "%s should be a valid kubetest config file")
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+	})
+
+	It("Should create and upgrade a workload cluster and run kubetest", func() {
```

Comment:
is this test a combination of https://github.com/kubernetes-sigs/cluster-api/blob/master/test/e2e/kcp_upgrade.go + https://github.com/kubernetes-sigs/cluster-api/blob/master/test/e2e/md_upgrades.go +  https://github.com/kubernetes-sigs/cluster-api/blob/master/test/e2e/machine_pool.go#L111 + https://github.com/kubernetes-sigs/cluster-api/blob/master/test/e2e/k8s_conformance.go ?

Reply:
Yes it creates a workload cluster, upgrades it and then runs conformance to verify the upgrade. The purpose is to eventually setup a job which tests the supported k8s versions for each release line. 

Check this discussion around https://github.com/kubernetes-sigs/cluster-api/issues/3693#issuecomment-730375355